File: Supplementary_webpages/static/gallery_res/Convex Optimization page 384/convex_optimization_384.cpp
/*
y_i = a_iᵀ x + w_i
x̂ = (∑_i a_i a_iᵀ)⁻¹ ∑_i y_i a_i

where

a_i ∈ ℝ^m: the measurement vectors  
w_i ∈ ℝ: measurement noise 
x ∈ ℝ^m: original vector 
*/
#include <Eigen/Core>
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <iostream>
#include <set>

struct convex_optimization_384ResultType {
    Eigen::VectorXd y;
    Eigen::VectorXd x̂;
    convex_optimization_384ResultType(const Eigen::VectorXd & y,
               const Eigen::VectorXd & x̂)
    : y(y),
    x̂(x̂)
    {}
};

/**
 * convex_optimization_384
 *
 * @param a  the measurement vectors  
 * @param w  measurement noise 
 * @param x  original vector 
 * @return x̂
 */
convex_optimization_384ResultType convex_optimization_384(
    const std::vector<Eigen::VectorXd> & a,
    const std::vector<double> & w,
    const Eigen::VectorXd & x)
{
    const long dim_0 = w.size();
    const long m = a[0].rows();
    assert( a.size() == dim_0 );
    for( const auto& el : a ) {
        assert( el.size() == m );
    }
    assert( x.size() == m );

    Eigen::VectorXd y(dim_0);
    for( int i=1; i<=dim_0; i++){
        y[i-1] = (double)(a.at(i-1).transpose() * x) + w.at(i-1);
    }

    Eigen::MatrixXd sum_0 = Eigen::MatrixXd::Zero(m, m);
    for(int i=1; i<=a.size(); i++){
        sum_0 += a.at(i-1) * a.at(i-1).transpose();
    }
    Eigen::MatrixXd sum_1 = Eigen::MatrixXd::Zero(m, 1);
    for(int i=1; i<=y.size(); i++){
        sum_1 += y[i-1] * a.at(i-1);
    }
    Eigen::VectorXd x̂ = (sum_0).inverse() * sum_1;

    return convex_optimization_384ResultType(y, x̂);
}


void generateRandomData(std::vector<Eigen::VectorXd> & a,
    std::vector<double> & w,
    Eigen::VectorXd & x)
{
    const int dim_0 = rand()%10;
    const int m = rand()%10;
    a.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        a[i] = Eigen::VectorXd::Random(m);
    }
    w.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        w[i] = rand() % 10;
    }
    x = Eigen::VectorXd::Random(m);
}


int main(int argc, char *argv[])
{
    srand((int)time(NULL));
    std::vector<Eigen::VectorXd> a;
    std::vector<double> w;
    Eigen::VectorXd x;
    generateRandomData(a, w, x);
    convex_optimization_384ResultType func_value = convex_optimization_384(a, w, x);
    std::cout<<"return value:\n"<<func_value.x̂<<std::endl;
    return 0;
}

File: Supplementary_webpages/static/gallery_res/Convex Optimization page 384/convex_optimization_384.py
"""
y_i = a_iᵀ x + w_i
x̂ = (∑_i a_i a_iᵀ)⁻¹ ∑_i y_i a_i

where

a_i ∈ ℝ^m: the measurement vectors  
w_i ∈ ℝ: measurement noise 
x ∈ ℝ^m: original vector 
"""
import numpy as np
import scipy
import scipy.linalg
from scipy import sparse
from scipy.integrate import quad
from scipy.optimize import minimize


class convex_optimization_384ResultType:
    def __init__( self, y, x̂):
        self.y = y
        self.x̂ = x̂


def convex_optimization_384(a, w, x):
    """
    :param :a : the measurement vectors  
    :param :w : measurement noise 
    :param :x : original vector 
    """
    a = np.asarray(a, dtype=np.float64)
    w = np.asarray(w, dtype=np.float64)
    x = np.asarray(x, dtype=np.float64)

    dim_0 = w.shape[0]
    m = a.shape[1]
    assert a.shape == (dim_0, m, )
    assert w.shape == (dim_0,)
    assert x.shape == (m,)

    y = np.zeros(dim_0)
    for i in range(1, dim_0+1):
        y[i-1] = (a[i-1].T.reshape(1, m) @ x).item() + w[i-1]
    sum_0 = np.zeros((m, m))
    for i in range(1, len(a)+1):
        sum_0 += (a[i-1]).reshape(m, 1) @ a[i-1].T.reshape(1, m)
    sum_1 = np.zeros((m, ))
    for i in range(1, len(y)+1):
        sum_1 += y[i-1] * a[i-1]
    x̂ = np.linalg.inv((sum_0)) @ sum_1
    return convex_optimization_384ResultType(y, x̂)


def generateRandomData():
    dim_0 = np.random.randint(10)
    m = np.random.randint(10)
    a = np.random.randn(dim_0, m, )
    w = np.random.randn(dim_0)
    x = np.random.randn(m)
    return a, w, x


if __name__ == '__main__':
    a, w, x = generateRandomData()
    print("a:", a)
    print("w:", w)
    print("x:", x)
    func_value = convex_optimization_384(a, w, x)
    print("return value: ", func_value.x̂)